### DISCUSSION TITLE
Challenges with using the Mina Fungible Token Standard for zkUSD

### DISCUSSION THREAD

**Discussion by mackcee** [zkUSD-Protocol/zkUSD#4] [General]
### Context

With developing zkUSD, a private, algorithmic stablecoin on the Mina Protocol. While the intention is to integrate the Mina Fungible Token (MFT) standard, technical limitations have prevented us from doing so in its current form.

### The Challenge

The primary obstacle is the Account Update Limit, which caps the number of updates in a transaction. zkUSD's minting process relies on additional account updates for decentralised price feeds, and a vault assertion flag, pushing us beyond the allowed limit with the standard. 

As explained by @dfstio in [this discord discussion](https://discord.com/channels/484437221055922177/1282303057526263879):

"given the Mina protocol limitation of 5 AccountUpdates with proof authorisation in one transaction and the fact that each token account AccountUpdate should be approved?

The maximum number of AccountUpdates with proof authorisation can be calculated using the formula

```
10.26*np + 10.08*n2 + 9.14*n1 < 69.45
where np: number of single proof updates
n2: number of pairs of signed/no-auth update
n1: number of single signed/no-auth update
```

`(69.45 - 9.14 (single feePayer))/10.26 = 5.87`, i.e max 5 proof AccountUpdates"

The vanilla standard mint function introduces the following account updates:

```javascript
  {
    label: 'feePayer',
    publicKey: '..TcKd',
    fee: '100000000',
    nonce: '0',
    authorization: '..Kx5X'
  },
  {
    label: 'FungibleToken.mint()',
    publicKey: '..vL9v',
    events: '[["2","1029302657826637044695563859193822697491331411233231028843717593204836076799","0","1000000000"]]',
    preconditions: {
      account: '{"state":[null,"25485543189759641532785121947742871732151404791418730633289207921913708665597","1","0",null,null,null,null]}'
    },
    mayUseToken: { parentsOwnToken: false, inheritFromParent: false },
    authorizationKind: {
      isSigned: false,
      isProved: true,
      verificationKeyHash: '25084457276132306637089336910977939820654927814172888515895248592725736067489'
    },
    authorization: undefined
  },
  {
    label: 'FungibleTokenAdmin.canMint()',
    publicKey: '..9nv9',
    callDepth: 1,
    preconditions: {
      account: '{"state":["25485543189759641532785121947742871732151404791418730633289207921913708665597","1",null,null,null,null,null,null]}'
    },
    mayUseToken: { parentsOwnToken: false, inheritFromParent: false },
    authorizationKind: {
      isSigned: false,
      isProved: true,
      verificationKeyHash: '4429719044247716591520838954939938661722471933627969303942407222910699643596'
    },
    authorization: undefined
  },
  {
    label: 'FungibleTokenAdmin.canMint() > AccountUpdate.createSigned()',
    publicKey: '..9nv9',
    incrementNonce: true,
    callDepth: 2,
    preconditions: { account: '{"nonce":{"lower":"1","upper":"1"}}' },
    useFullCommitment: true,
    mayUseToken: { parentsOwnToken: false, inheritFromParent: false },
    authorizationKind: { isSigned: true, isProved: false },
    authorization: { signature: '..uhEh' }
  },
  {
    label: 'FungibleToken.mint().token.mint()',
    publicKey: '..TcKd',
    tokenId: '..EWCa',
    balanceChange: { magnitude: '1000000000', sgn: 'Positive' },
    callDepth: 1,
    mayUseToken: { parentsOwnToken: true, inheritFromParent: false }
  },
  {
    label: 'FungibleToken.mint() > AccountUpdate.create()',
    publicKey: '..vL9v',
    tokenId: '..EWCa',
    balanceChange: { magnitude: '1000000000', sgn: 'Positive' },
    callDepth: 1,
    mayUseToken: { parentsOwnToken: true, inheritFromParent: false }
  }
```

At a minimum this calculates to:

```
1x FeePayer (single signed):       09.14
2x Proven account updates:         20.52
3x Single signed/no-auth update:   27.42
----------------------------------------
Total cost:                        57.08
```

The zkUSD tokens minting functionality has to be locked down to be only callable from within a zkUSD Vault, in order to do this we implement the following pattern as detailed by @mitschabaude here: https://github.com/o1-labs/o1js/blob/main/src/examples/zkapps/joint-update.ts

This pattern introduces one more additional proven update which puts us to:

```
Vanilla Mint:                       57,08
Assertion flag on calling contract: 10.26
-----------------------------------------
Total cost:                         67,34
```

As you can see we are at the limit of `69.45` for the amount of account updates we can have in the transaction. 

### Decentralised on-chain price feed

In order to mint zkUSD, we need to work out what the health factor of the vault would be after the mint. To do this, we need to use the price to calculate the value of the collateral currently locked within the vault. For the system to be decentralised, we need to get this from an on-chain oracle contract. However this creates one additional account update as we get the price from the oracle contract, pushing us over the edge

```javascript
      {
        label: 'ZkUsdPriceFeedOracle.getPrice()',
        publicKey: '..Aaf9',
        callDepth: 1,
        preconditions: {
          network: '{"blockchainLength":{"lower":"0","upper":"0"}}',
          account: '{"state":["1000000000",null,null,null,null,"0",null,null]}'
        },
        mayUseToken: { parentsOwnToken: false, inheritFromParent: false },
        authorizationKind: {
          isSigned: false,
          isProved: true,
          verificationKeyHash: '3392518251768960475377392625298437850623664973002200885669375116181514017494'
        },
        authorization: undefined
      },
```

### Design Trade-Off: Decentralised Price Feed vs. MFT Standard

The core trade-off we faced revolved around prioritising either:
1. **Decentralisation of the Price Feed**: A decentralised oracle contract ensures zkUSD remains trustless and robust by fetching on-chain price data from a distributed network of oracles.
2. **MFT Standard Integration**: Utilising the MFT standard provides maximum interoperability and adherence to existing protocols within the Mina ecosystem.

**Why Decentralisation Took Precedence**

- A decentralised price feed is critical for zkUSD's integrity and security, particularly given the protocol's role in managing potentially significant amounts of locked collateral. Without a decentralised feed, the system would rely on centralised components, introducing unacceptable risks.

- While the Mina Fungible Token (MFT) standard offers significant benefits in terms of interoperability, integrating it would have added an extra account update due to its reliance on an external administrative contract. This conflicted with the system's requirements and pushed the transaction beyond the Account Update Limit.

- A Trade-Off for Interoperability:
  - By deviating from the MFT standard, there is a cost to interoperability, as any application wishing to integrate zkUSD will need to interact directly with the custom token contract code. This introduces an extra step for developers.
  - However, this trade-off is deemed acceptable given the critical importance of maintaining a trustless, decentralised price feed. Additionally, zkUSD’s role as a stablecoin—a foundational DeFi primitive—means it is likely to attract strong adoption and incentivise applications to take the extra steps to integrate with its token contract.

In light of these considerations, prioritising decentralisation over strict adherence to the MFT standard ensures zkUSD remains secure, robust, and aligned with its long-term mission of being a stable and reliable asset within the Mina ecosystem.

### Open Discussion
We would love to gather feedback from the Mina community on the following:

1. Network-Level Adjustments:

- Would increasing the Account Update Limit create more room for innovation without compromising performance?

2. Trade-Offs:

- Do you agree with prioritising decentralisation over strict adherence to the MFT standard?
- Are there other potential solutions we might have overlooked?

We welcome all feedback and suggestions to help refine zkUSD and contribute to the broader Mina ecosystem.

**Comment by dfstio** [comment]
You can resolve it by defining your custom FungibleTokenAdmin that is allowed by the Fungible Token standard and defining it in a way that replaces two AccountUpdates consumed now by FungibleTokenAdmin with one proof update.
You can replace the admin signature with proof of something in your custom admin.

An example of such an admin that does whitelist check, and how it is used together with FungibleToken, you can see at
https://github.com/zkcloudworker/zkcloudworker-lib/blob/main/src/mina/token/FungibleTokenWhitelistedAdmin.ts
https://github.com/zkcloudworker/zkcloudworker-lib/blob/main/src/mina/token/token.ts#L38

You can replace the whitelisting logic with your logic that checks price feeds or keep it to have whitelisting functionality and just add price feeds. In the best case scenario, you will have one AccountUpdate left for something else.

**Reply by mackcee** [reply]
We actually followed this pattern in a previous commit: https://github.com/zkUSD-Protocol/zkUSD/tree/7f32dc2bb323e6a1cd41c318258079b5cea9695c/src

The trouble is that we can't combine those two account updates into one because we have to use `interactionFlag` from the calling vault contract. This is to ensure that the mint is only called from within the `mintZkUsd()` function of an instance of a zkUsdVault contract.
Therefore we will always have two account updates, the `FungibleTokenAdmin.canMint()` and the `ZkUsdVault.assertInteractionFlag()`

On that above commit: 

We are getting the price from the `getPrice()` function in the oracle contract. We were making a mistake in that commit by not adding the @method decorator to the `getPrice()` function.  The @method decorate is required to ensure that the price data is actually the data on chain by creating an account update for the oracle contract and adding that as a precondition inside it.

Right now, by absorbing the administrative functionality into the token contract, we get rid of the `FungibleTokenAdmin.canMint()` which puts us just under the limit.

**Reply by dfstio** [reply]
Great presentation today, thank you!

To solve the issue with the `interactionFlag`, you can consider installing the wallet contracts to the token accounts of the FungibleTokenAdmin. It will look like this:

- The main contract is standard FungibleToken that uses ZkUsdFungibleTokenAdmin
- ZkUsdFungibleTokenAdmin implements the maximum functions possible, including keeping the record of the last prices and interacting with oracles to minimize the account updates needed for working with wallets
- ZkUsdProtocolVault is installed to the token accounts to the ZkUsdFungibleTokenAdmin. This way, it will only be possible to call the ZkUsdProtocolVault through ZkUsdFungibleTokenAdmin if you block the execution of the approveBase() function. You can see the example of the installing the contracts in the NFT standard:
https://github.com/dfstio/nft-standard-draft/blob/744548384057b61a8770bac520ad117b8f882f3f/src/contracts/collection.ts#L375

Therefore, in one call it will be possible to execute the code in ZkUsdFungibleTokenAdmin and ZkUsdProtocolVault. It does create several account updates, but probably the number of account updates will be less.

You can also remove the signatures that create signed AccountUpdates in cases when you're going short of account updates by replacing them with proofs that verify signatures. It will cause regression of the user experience, so it is better to do it only for administrative functions and include some custom nonce written in the contract state to protect from proof replaying.

**Reply by mackcee** [reply]
I've looked through the code and it is a really impressive solution!

In our case though, the minting can only occur once a vault has already been deployed and collateralised. Which is to say, that the vault has Mina deposited into it. Therefore the vault couldn't be installed to the token account of the admin contract, as it needs to be able to hold Mina and as far as I am aware, the vault needs to be of token id 1 to hold Mina. Please correct my understanding here if I am wrong?

However I still think I might be able to utilise a similar pattern by loading all the token interactions (minting / burning / liquidating) into the ZkUsdFungibleTokenAdmin and combining it with the oracle an protocol vault contracts. Let me play around with it and see what I can come up with.

**Reply by dfstio** [reply]
You can hold Mina in the Admin contract for all the vaults.

**Comment by 45930** [comment]
> Would increasing the Account Update Limit create more room for innovation without compromising performance?

No, it would compromise performance :) but o1Labs is working on the performance aspect.  We know we need to raise the AU limit for zkApps to flourish.

> Do you agree with prioritising decentralisation over strict adherence to the MFT standard?

I think the best thing is to view the project as experimental and keep both possibilities in mind.  I think you've already done a great job of seriously evaluating the standard and trying to implement it.  Your feedback has triggered a lot of good discussion, which is an amazing outcome!  If the standard doesn't progress to the point where you can use it, or the network doesn't upgrade fast enough, and you can make progress in another way, that is also fantastic.  But then if you commit to doing something custom and the standard gets a lot of traction, your project may not work with the rest of the ecosystem.  I would basically recommend not committing one way or the other just yet.

**Reply by mackcee** [reply]
Thanks for your input!

We have actually managed to get the standard working by implementing @dfstio's token account model. By including the price feed functionality in the admin contract, we can reduce the number of AU's needed and can squeeze everything in so we can still use the standard.

**Reply by 45930** [reply]
Glad to hear it!
